### PR TITLE
fix non numeric accumulate(op, v0, x) (#25506)

### LINKED ISSUE #25506: accumulate(op, v0, x) fails on arrays of non-numeric types
The following produces an error:

```julia
julia> accumulate((acc, x) -> acc+x[1], 0, [(1,2), (3,4), (5,6)])
ERROR: MethodError: no method matching rcum_promote_type(::##1#2, ::Type{Int64}, ::Type{Tuple{Int64,Int64}})
Closest candidates are:
  rcum_promote_type(::Any, ::Type{T<:Number}) where T<:Number at multidimensional.jl:585
  rcum_promote_type(::Any, ::Type{T}) where T at multidimensional.jl:586
  rcum_promote_type(::Any, ::Type{T}, !Matched::Type{S<:Number}) where {T, S<:Number} at multidimensional.jl:584
  ...
Stacktrace:
 [1] accumulate at ./multidimensional.jl:760 [inlined] (repeats 2 times)
 [2] macro expansion at ./REPL.jl:97 [inlined]
 [3] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
```
The same happens for
```julia
accumulate((acc, x) -> x, "", ['a', 'b', 'c'])
```

It seems that `accumulate(op, v0, x)` tries to guess the type of the output array based on both the type of the initial value and the type of the input array, while it should only consider the first.

Changing this line:
https://github.com/JuliaLang/julia/blob/4440b0f2a12479dfc691c6b4d079ba113290d7de/base/multidimensional.jl#L1105
to:
```julia
    T = rcum_promote_type(op, typeof(v0))
```
seems to fix the issue, but I am not sure whether there is a reason to prefer the original version.

### PR DESCRIPTION
**Pull request by jw3126**
Fix #25506

### CODE REVIEW

**Comment by jw3126** [comment]
@simonbyrne You said in the issue   #25506:

> A full solution to this would probably involve extending #25051 to accumulate

Can you explain whats wrong with the fix in this PR and provide an example that fails?

**Comment by jw3126** [comment]
@simonbyrne I think the `accumulate` case is actually simpler then `(map)reduce`. Because
* accumulate over empty collection is empty, we don't need to come up with an element
* reduce does a lot of widening, which accumulate does not

**Comment by simonbyrne** [comment]
> reduce does a lot of widening, which accumulate does not

I would like to change that to be consistent

**Comment by jw3126** [comment]
Its was a conscious design decision, that `accumulate` does not widen.

**Comment by simonbyrne** [comment]
I'm actually trying to find that at the moment: do you have a link?

I would like to change `accumulate` to use the same machinery as `reduce`, so that `accumulate(+,...)` won't widen, but `cumsum(...)` will.

**Comment by jw3126** [comment]
Okay #18364 for `cumsum` and a bit in #18931.

**Comment by jw3126** [comment]
FWIW I am also in favor of making this consistent with `reduce`.

**Comment by simonbyrne** [comment]
Thanks for the links, I'll leave that to a separate issue.

Even if we don't widen, I would still prefer to use `reduce_first` instead of defining `rcum_convert`.

File: base/multidimensional.jl [hunk 1]
```diff
@@ -723,14 +712,14 @@ function accumulate_pairwise!(op::Op, result::AbstractVector, v::AbstractVector)
...
-    @inbounds result[i1] = v1 = v[i1]
+    @inbounds result[i1] = v1 = reduce_first(op,v[i1])
     n == 1 && return result
     _accumulate_pairwise!(op, result, v, v1, i1+1, n-1)
     return result
 end
 
 function accumulate_pairwise(op, v::AbstractVector{T}) where T
-    out = similar(v, rcum_promote_type(op, T))
+    out = similar(v, promote_op(op, T))
```
Comment (simonbyrne) [inline base/multidimensional.jl]: This (and others) should probably be `promote_op(op, T, T)`: though built-in operators provide unary versions, anonymous functions typically won't.

Comment (jw3126) [reply base/multidimensional.jl]: Oh thanks! I somehow assumed `promote_op(op,T) = promote_op(op,T,T)` without checking. Documenting it is a good idea!

**Review by simonbyrne — APPROVED** [review]

**Comment by jw3126** [comment]
@martinholters What do you mean by empty case? Are you talking about `accumulate(op, v0, iter)` vs `accumulate(op, iter)` or about `iter` being empty? I am not familiar with the broadcast code, can you explain the return type mechanism there or link to some relevant lines?

**Comment by martinholters** [comment]
> What do you mean by empty case?

The situation where the output is empty, so where `iter` is empty. If `v0` is present it might be better to us its type as the eltype instead of calling `promote_op`, not sure.

> I am not familiar with the broadcast code, can you explain the return type mechanism there

Conceptually, it works as follows:
1) Compute first output element, allocate output container with first element's type as eltype, store first element.
2) Compute next element. If its type is a subtype of the containers eltype, store it. Otherwise, `typejoin` the current eltype and the type of the newly computed element, allocate new container with the new eltype, copy previous work, store new element.
3) Repeat at 2) until all elements have been computed.

If the code is type stable, inference can figure out that the type check in 2) can be optimized away. So a sufficiently tweaked implementation can be quite efficient.

However,
```julia
julia> accumulate(*, ['a', 'b'])
2-element Array{Any,1}:
 'a' 
 "ab"
```
may not be what you want in the end. (But then again, `accumulate(*, "", ['a', 'b'])` would give the desired `Vector{String}`.)

**Comment by simonbyrne** [comment]
To save you some time hunting around (the broadcast code is a bit confusing): `broadcast` [calls `Base._return_type`](https://github.com/JuliaLang/julia/blob/4873dc007cf4a837c7e7349a6a9d70c2339402d6/base/broadcast.jl#L537-L538): if it is a concrete type, this is used to [create an output array](https://github.com/JuliaLang/julia/blob/4873dc007cf4a837c7e7349a6a9d70c2339402d6/base/broadcast.jl#L628), otherwise it [checks if empty and determines the initial element](https://github.com/JuliaLang/julia/blob/4873dc007cf4a837c7e7349a6a9d70c2339402d6/base/broadcast.jl#L640-L657), then [uses the iterative procedure discussed above](https://github.com/JuliaLang/julia/blob/4873dc007cf4a837c7e7349a6a9d70c2339402d6/base/broadcast.jl#L505-L516).

**Comment by jw3126** [comment]
Mhhh, seeing
```julia
julia> accumulate(*, ['a', 'b'])
2-element Array{Any,1}:
 'a' 
 "ab"
```
hurts my eyes. Getting `accumulate(*, ['a', 'b'])` right, was one of the motivating examples of this PR :smile:.

**Comment by simonbyrne** [comment]
I think we should still be able to use `reduce_first` with this idea.

**Comment by jw3126** [comment]
@simonbyrne okay you are right, we could still use `reduce_first`. I am still not convinced, this adds so much complexity to `accumulate`. What would be an example, that goes haywire with this PR but can be handled gracefully by the `broadcast` approach?

**Comment by martinholters** [comment]
> Getting `accumulate(*, ['a', 'b'])` right, was one of the motivating examples of this PR

Maybe we have different opinions of what "right" is here? To me, `accumulate(*, ['a', 'b']) == ['a', 'a'*'b']` looks pretty right. Anything else has lot of DWIM to it. Compare with
```julia
julia> accumulate(*, [1m, 1m]) # using Unitful (result faked)
2-element Array{Unitful.Quantity{Int64,D,U} where U where D,1}:
   1 m
 1 m^2
```
which is basically the same situation. Would we want to enforce some other element type (and likely cause a failure by doing so)?

**Comment by StefanKarpinski** [comment]
To me the difference between `reduce(*, ['a', 'b'])` and `reduce(*, [1m, 1m])` is simply that `'a'` and `"ab"` should probably promote to `String` as common type – we don't currently do that, which is something we could easily change. On the other hand, `1m` and `1m^2` should certainly not promote to a common concrete type (what unit would it be?). So perhaps a side fix here would be adding sufficiently general versions of these methods:
```jl
Base.promote_rule(::Type{Char}, ::Type{S}) where S<:AbstractString = S
Base.convert(::Type{String}, c::Char) = string(c)
```
allowing this to work:
```jl
julia> promote('a', "ab")
("a", "ab")
```
That would, I believe, naturally make the `reduce(*, ['a', 'b'])` example act as desired. Note that we were once reluctant to be too fast and loose with character <=> string conversions since characters were subtypes of integers as code points, but that's no longer the case, so I think it's safe.

**Comment by jw3126** [comment]
Okay maybe the issue is indeed more with general `Char` <=> `String` conversion policy, then with `accumulate`. I still think, `accumulate(*, ['a']) = ["a"]` and `reduce(*, ['a']) = "a"` are the behavior we should aim for. 
But it agree it is not ideal to hack that into `reduce` and `accumulate` and it should instead come out of the general conversion machinery.

**Comment by StefanKarpinski** [comment]
Given that `'a'*'b' === "ab"` for "multiplicative purposes, characters are embedded into strings as one-character strings. Therefore, I think that when you ask what the multiplicative identity of characters is, the only reasonable answer is that it's the multiplicative identity for strings, i.e. `""`.

**Comment by simonbyrne** [comment]
I agree, which is why I defined a `reduce_empty` method for it in #25051. My point was that it isn't well-defined for all possible operations/types.

**Comment by martinholters** [comment]
Assuming #25728 lands, could this serve as a prototype implementation?
```julia
function accumulate(f, x)
    res = Vector{Any}(uninitialized, length(x))
    res[1] = x[1]
    for i in 2:length(x)
        res[i] = f(res[i-1], x[i])
    end
    restype = mapreduce(typeof, promote_type, res)
    return convert(Vector{restype}, res)
end
```
Obviously, this is not the most efficient way to go about it and is restricted to non-empty `Vector` inputs, but would it produce the results we want?

**Comment by simonbyrne** [comment]
I've created a WIP fix in #25766, which I think might be a more general solution.

**Comment by jw3126** [comment]
I close this in favor of #25766.

**Pull request closed without merging**